multicluster/controllers/multicluster/stale_controller.go

```diff
@@ -102,128 +122,169 @@ func (c *StaleController) cleanupStaleServiceResources(remoteCluster commonarea.
+		if err := c.Client.Delete(ctx, &ciImp, &client.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
+			return err
+		}
+	}
+	return nil
+}
+
+// cleanupServiceResourceExport removes any Service/Endpoint kind of ResourceExports when there is no
+// corresponding ServiceExport in the local cluster.
+func (c *StaleResCleanupController) cleanupServiceResourceExport(commonArea commonarea.RemoteCommonArea,
+	resExpList *mcsv1alpha1.ResourceExportList) error {
 	svcExpList := &k8smcsv1alpha1.ServiceExportList{}
 	if err := c.List(ctx, svcExpList, &client.ListOptions{}); err != nil {
 		return err
 	}
-	resExpItems := resExpList.Items
+	allResExpItems := resExpList.Items
 	svcExpItems := svcExpList.Items
+	var staleResExpItems []mcsv1alpha1.ResourceExport
 
-	for k, re := range resExpItems {
+	for i, resExp := range allResExpItems {
+		if (resExp.Spec.Kind == common.ServiceKind || resExp.Spec.Kind == common.EndpointsKind) && resExp.Labels[common.SourceClusterID] == c.localClusterID {
+			staleResExpItems = append(staleResExpItems, allResExpItems[i])
+		}
+	}
+
+	for k, re := range staleResExpItems {
 		for _, se := range svcExpItems {
 			if re.Spec.Name == se.Name && re.Spec.Namespace == se.Namespace {
 				// Set the valid ResourceExport item as empty ResourceExport, then all left non-empty items should be removed.
-				resExpItems[k] = mcsv1alpha1.ResourceExport{}
+				staleResExpItems[k] = mcsv1alpha1.ResourceExport{}
 			}
 		}
 	}
 
-	for _, r := range resExpItems {
+	for _, r := range staleResExpItems {
 		re := r
 		if re.Name != "" {
 			klog.InfoS("Cleaning up ResourceExport", "ResourceExport", klog.KObj(&re))
-			if err := remoteCluster.Delete(ctx, &re, &client.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
+			if err := commonArea.Delete(ctx, &re, &client.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
 				return err
 			}
 		}
 	}
 	return nil
 }
 
-func (c *StaleController) cleanupACNPResources(resImpList *mcsv1alpha1.ResourceImportList) error {
-	acnpList := &crdv1alpha1.ClusterNetworkPolicyList{}
-	if err := c.List(ctx, acnpList, &client.ListOptions{}); err != nil {
+// cleanupClusterInfoResourceExport removes any Gateway kind of ResourceExports when there is no
+// Gateway in the local cluster.
+func (c *StaleResCleanupController) cleanupClusterInfoResourceExport(commonArea commonarea.RemoteCommonArea,
+	resExpList *mcsv1alpha1.ResourceExportList) error {
+	var gws mcsv1alpha1.GatewayList
+	if err := c.Client.List(ctx, &gws, &client.ListOptions{}); err != nil {
 		return err
 	}
-	staleMCACNPItems := map[string]crdv1alpha1.ClusterNetworkPolicy{}
-	for _, acnp := range acnpList.Items {
-		if _, ok := acnp.Annotations[common.AntreaMCACNPAnnotation]; ok {
-			staleMCACNPItems[acnp.Name] = acnp
+
+	if len(gws.Items) == 0 {
+		ciExport := &mcsv1alpha1.ResourceExport{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: commonArea.GetNamespace(),
+				Name:      c.localClusterID + "-clusterinfo",
```

Comment:
should call `newClusterInfoResourceExportName`

Reply:
rename is as `newCIResourceExportName`

Reply:
I mean it should call the util function `newClusterInfoResourceExportName` like other code, instead of keeping a magic string here.

I have no problem with the previous variable name, and the new name is confusing as it's not a "name" but a "resourceExport".

Reply:
oh, yes, I misread it and thought it's a NamespacedName, lost my mind..., fixing